controller/contour/controller.go

```diff
@@ -0,0 +1,123 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package contour
+
+import (
+	"context"
+	"fmt"
+	"k8s.io/apimachinery/pkg/types"
+
+	retryable "github.com/projectcontour/contour-operator/util/retryableerror"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/projectcontour/contour-operator/util/slice"
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	operatorv1alpha1 "github.com/projectcontour/contour-operator/api/v1alpha1"
+)
+
+const (
+	contourFinalizer string = "contour.operator.projectcontour.io/finalizer"
+
+	defaultContourNamespace = "projectcontour"
+)
+
+// Config holds all the things necessary for the controller to run.
+type Config struct {
+	// Image is the name of the Contour container image.
+	Image string
+}
+
+// Reconciler reconciles a Contour object.
+type Reconciler struct {
+	Config
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=operator.projectcontour.io,resources=contours,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=operator.projectcontour.io,resources=contours/status,verbs=get;update;patch
+
+func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	_ = r.Log.WithValues("contour", req.NamespacedName)
+
+	r.Log.Info("reconciling", "request", req)
+
+	// Only proceed if we can get the state of contour.
+	contour := &operatorv1alpha1.Contour{}
+	if err := r.Get(context.TODO(), req.NamespacedName, contour); err != nil {
+		if errors.IsNotFound(err) {
+			// This means the contour was already deleted/finalized and there are
+			// stale queue entries (or something edge triggering from a related
+			// resource that got deleted async).
+			r.Log.Info("contour not found; reconciliation will be skipped", "request", req)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get contour %q: %v", req, err)
+	}
+
+	// The contour is safe to process, so ensure current state matches desired state.
+	desired := true
```

Comment:
I guess it's ok for now, but 

1. passing a bool is harder to read and understand than an enum
2. teardown probably needs to be done in the reverse order

Reply:
Yeah, I refactored based on https://github.com/projectcontour/contour-operator/pull/10#discussion_r481492143.